fix(vscode): toImportSpecifier() prepends current Windows drive to POSIX absolute paths

- Markuplint version: `4.18.1` (v4 branch HEAD)
- Parser lang: N/A (vscode extension server)
- Node.js version: `24` (CI matrix)
- OS: Windows

## Describe the bug

`vscode/src/server/get-module.ts`'s `toImportSpecifier()` prepends the current Windows drive letter to POSIX absolute paths. When called on a Windows runtime with `/tmp/foo.mjs`, it returns `'file:///D:/tmp/foo.mjs'` instead of the expected `'file:///tmp/foo.mjs'`. POSIX paths must round-trip unchanged because the VS Code extension can receive POSIX paths from a Windows process via WSL or remote development scenarios.

The unit tests at `vscode/src/server/get-module.spec.ts:39` and `:43` already encode this contract, and they fail on Windows CI.

## Code Example or Playground URL

```ts
// vscode/src/server/get-module.spec.ts
test('converts a POSIX absolute path to a file:// URL', () => {
  expect(toImportSpecifier('/tmp/markuplint/lib/index.mjs'))
    .toBe('file:///tmp/markuplint/lib/index.mjs');
});

test('URL-encodes spaces in POSIX paths', () => {
  expect(toImportSpecifier('/tmp/my project/lib.mjs'))
    .toBe('file:///tmp/my%20project/lib.mjs');
});
```

## Steps To Reproduce

1. `git checkout v4`
2. Run the test on Windows: `npx vitest run vscode/src/server/get-module.spec.ts`
3. Both POSIX-path tests fail with received `'file:///D:/tmp/...'`

## The current behavior

```
AssertionError: expected 'file:///D:/tmp/markuplint/lib/index.mjs'
                to be 'file:///tmp/markuplint/lib/index.mjs'
```

The implementation likely uses `pathToFileURL()` (or equivalent), which on Windows resolves `/tmp/...` against the current drive (`D:` on the GitHub-hosted Windows runner) and prepends the drive letter.

## The expected behavior

POSIX absolute paths must round-trip: `/tmp/foo.mjs` → `file:///tmp/foo.mjs`, regardless of the host OS.

## Root cause and history

- Introduced by **#3807** (`fix(vscode): convert absolute paths to file:// URLs before import`, merged 2026-04-24).
- That PR only changed files under `vscode/`. The test matrix in `.github/workflows/test.yml` requires `paths: packages/**` (and a few siblings), so the matrix was **not triggered** when #3807 itself was merged — the latent Windows bug slipped through review.
- First exposed on **2026-04-29** when an unrelated v4-base PR (#3788, renovate jsdom) was rebased onto v4 and ran the full matrix.
- Reproduced on 2026-05-04, 2026-05-07, and 2026-05-09 across multiple unrelated v4-base PRs.

CI runs reproducing the failure:

- https://github.com/markuplint/markuplint/actions/runs/25109268035 (2026-04-29, #3788 after rebase)
- https://github.com/markuplint/markuplint/actions/runs/25297742579 (2026-05-04, #3788)
- https://github.com/markuplint/markuplint/actions/runs/25489791893 (2026-05-07, #3788)
- https://github.com/markuplint/markuplint/actions/runs/25593369766 (2026-05-09, #3834)

CI run **before** the regression, confirming the path used to work:

- https://github.com/markuplint/markuplint/actions/runs/24769012803 (2026-04-22, #3788, pre-#3807)

## Impact

- All v4-base PRs that touch `packages/**` are blocked from merging by this Windows-only test failure.
- #3788 (renovate `jsdom@29`) — open since 2026-04-22, blocked.
- #3834 (#3823 astro-parser fix) — blocked.

## Suggested fix direction

1. Have `toImportSpecifier()` short-circuit on a leading `/`: if the input is a POSIX-style absolute path, prepend `file://` directly without going through `pathToFileURL()`.
2. Alternatively, narrow the Windows path detection (e.g. `path.isAbsolute` + drive-letter regex) so `pathToFileURL()` is only called for true Windows-style absolute paths.
3. Add a regression test that mocks `process.platform = 'win32'` so the bug is caught on POSIX CI as well — the existing tests rely on the runtime OS, which is why the bug was never caught locally.

## Process improvement

The `paths:` filter in `test.yml` excludes `vscode/**`. Consider either:

- Adding `vscode/**` to the trigger paths, OR
- Scoping `vitest run` to package paths so vscode tests run in a dedicated vscode workflow.

Either change would have caught this regression at the #3807 review stage.

--- a/vscode/src/server/get-module.spec.ts
+++ b/vscode/src/server/get-module.spec.ts
@@ -35,12 +35,28 @@ describe('toImportSpecifier', () => {
 		expect(toImportSpecifier('c:\\Users\\foo#bar\\lib.mjs')).toBe('file:///c:/Users/foo%23bar/lib.mjs');
 	});
 
+	test('URL-encodes reserved characters like ? in Windows paths', () => {
+		expect(toImportSpecifier('c:\\Users\\foo?bar\\lib.mjs')).toBe('file:///c:/Users/foo%3Fbar/lib.mjs');
+	});
+
 	test('converts a POSIX absolute path to a file:// URL', () => {
 		expect(toImportSpecifier('/tmp/markuplint/lib/index.mjs')).toBe('file:///tmp/markuplint/lib/index.mjs');
 	});
 
 	test('URL-encodes spaces in POSIX paths', () => {
 		expect(toImportSpecifier('/tmp/my project/lib.mjs')).toBe('file:///tmp/my%20project/lib.mjs');
+	});
+
+	test('URL-encodes non-ASCII characters in POSIX paths', () => {
+		expect(toImportSpecifier('/home/太郎/lib.mjs')).toBe('file:///home/%E5%A4%AA%E9%83%8E/lib.mjs');
+	});
+
+	test('URL-encodes reserved characters like # in POSIX paths', () => {
+		expect(toImportSpecifier('/home/foo#bar/lib.mjs')).toBe('file:///home/foo%23bar/lib.mjs');
+	});
+
+	test('URL-encodes reserved characters like ? in POSIX paths', () => {
+		expect(toImportSpecifier('/home/foo?bar/lib.mjs')).toBe('file:///home/foo%3Fbar/lib.mjs');
 	});
 
 	test('does not convert bare module specifiers like "markuplint"', () => {

--- a/vscode/src/server/get-module.spec.ts
+++ b/vscode/src/server/get-module.spec.ts
@@ -36,7 +36,7 @@ describe('toImportSpecifier', () => {
 	});
 
 	test('URL-encodes reserved characters like ? in Windows paths', () => {
-		expect(toImportSpecifier('c:\\Users\\foo?bar\\lib.mjs')).toBe('file:///c:/Users/foo%3Fbar/lib.mjs');
+		expect(toImportSpecifier('c:\\Users\\foo?\\lib.mjs')).toBe('file:///c:/Users/foo%3F/lib.mjs');
 	});
 
 	test('converts a POSIX absolute path to a file:// URL', () => {
@@ -56,7 +56,7 @@ describe('toImportSpecifier', () => {
 	});
 
 	test('URL-encodes reserved characters like ? in POSIX paths', () => {
-		expect(toImportSpecifier('/home/foo?bar/lib.mjs')).toBe('file:///home/foo%3Fbar/lib.mjs');
+		expect(toImportSpecifier('/home/foo?/lib.mjs')).toBe('file:///home/foo%3F/lib.mjs');
 	});
 
 	test('does not convert bare module specifiers like "markuplint"', () => {

--- a/vscode/src/server/get-module.ts
+++ b/vscode/src/server/get-module.ts
@@ -2,7 +2,6 @@ import type { Log } from '../types.js';
 import type { ARIAVersion } from '@markuplint/ml-spec';
 
 import path from 'node:path';
-import { pathToFileURL } from 'node:url';
 
 import { Files } from 'vscode-languageserver/node.js';
 
@@ -18,7 +17,10 @@ import { Files } from 'vscode-languageserver/node.js';
  * code path.
  *
  * Mirrors `packages/@markuplint/file-resolver/src/general-import.ts` —
- * keep the two in sync when adjusting Windows-path handling.
+ * keep the two in sync when adjusting Windows-path handling. Note:
+ * file-resolver currently still relies on `pathToFileURL()` and carries
+ * the same Windows / POSIX-absolute mismatch; tracked separately as
+ * #3840 (dev / v5).
  *
  * @param modPath - A module specifier — typically the result of `Files.resolve`
  *   or `require.resolve`, which may be a bare module name (`markuplint`), a
@@ -36,19 +38,28 @@ export function toImportSpecifier(modPath: string): string {
 	if (!isWindowsAbsolute && !isPosixAbsolute) {
 		return modPath;
 	}
+	// Build the `file://` URL explicitly for both branches. Node's
+	// `pathToFileURL()` is intentionally avoided: on Windows it resolves a
+	// POSIX-style absolute path against the *current drive* and emits
+	// `file:///D:/tmp/foo` instead of `file:///tmp/foo` (#3836); on POSIX
+	// it likewise mishandles Windows-style drive paths. Constructing the
+	// URL ourselves keeps the function OS-independent so POSIX CI exercises
+	// the Windows code path. Each segment is percent-encoded so that
+	// spaces (`Program Files`), non-ASCII characters (e.g. Japanese
+	// usernames), and URL-reserved characters like `#` / `?` do not get
+	// reinterpreted as fragment/query delimiters by Node's URL parser.
 	if (isWindowsAbsolute) {
-		// pathToFileURL on a POSIX runtime misinterprets Windows-style paths,
-		// so build the URL explicitly. Percent-encode each path segment so that
-		// spaces (`Program Files`), non-ASCII characters (e.g. Japanese
-		// usernames), and URL-reserved characters like `#` / `?` do not get
-		// reinterpreted as fragment/query delimiters by Node's URL parser.
 		// The drive-letter segment (`c:`) is kept as-is to match the
 		// `pathToFileURL` output shape on Windows (`file:///c:/...`).
 		const [drive, ...rest] = modPath.replaceAll('\\', '/').split('/');
 		const encoded = [drive, ...rest.map(segment => encodeURIComponent(segment))].join('/');
 		return `file:///${encoded}`;
 	}
-	return pathToFileURL(modPath).href;
+	// POSIX absolute path. Splitting `/tmp/foo` by `/` yields `['', 'tmp',
+	// 'foo']`; the leading empty element produces the `file:///` prefix
+	// after `join('/')`, so the round-trip is exactly `file:///tmp/foo`.
+	const segments = modPath.split('/').map(segment => encodeURIComponent(segment));
+	return `file://${segments.join('/')}`;
 }
 
 export async function getModule(log: Log): Promise<Module> {

